### DISCUSSION TITLE
[CT-3428] [Feature] Allow usage of macros in model configuration defined in dbt_project.yaml

### DISCUSSION THREAD

**Discussion by rlh1994** [dbt-labs/dbt-core#9172] [Ideas]
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

Currently it is possible to use a macro within the config block of model to make some of this configuration dynamic, such as based on a target or variable (see example [here](https://github.com/snowplow/dbt-snowplow-unified/blob/d7365445f5a7fd345dc083881d6485ad9439cb7e/models/base/manifest/snowplow_unified_incremental_manifest.sql#L11)). If you try and use a macro directly within the `dbt_project.yml` however, you get an error.

I understand these are parsed at different times, but I would like to be able to use a macro within the project yaml for model configs.

### Describe alternatives you've considered

Manually adding the macro to each model file, but this is not possible for models in packages.

### Who will this benefit?

Any user who wishes to (somewhat) dynamically control configs across models.

### Anything else?

There was a previous discourse post about this here: https://discourse.getdbt.com/t/how-can-i-add-model-configuration-using-a-macro-on-this-in-dbt-project-yml/5910

**Comment by dbeatty10** [comment]
This is a big idea, so I'm going to upgrade this to a Discussion.

Here's some past discussion on related topics:
- https://github.com/dbt-labs/dbt-core/discussions/6170
- https://github.com/dbt-labs/dbt-core/discussions/6280

Quick context:
It's not a top priority for us to dedicate time to dive into the assessment details on this one right now. The reason we haven't done this to-date is due to performance reasons. This wouldn't be an easy lift, and and it would take time and energy to implement.

**Comment by rlh1994** [comment]
Just to add to this but for an even simpler example, it is not currently possible to use variables directly within the `models` section of the `dbt_project.yml` either, which means that it is not possible to, within this file, set grants based on variable (something that would be helpful for some of our larger packages!) I'm left to either pass it through all model configs individually, or rebuild a grants macro that supports multiple warehouses unfortunately...
